rfcs/45-stageless.md

```diff
@@ -0,0 +1,725 @@
+# Feature Name: `stageless`
+
+## Summary
+
+The existing stage boundaries result in a large number of complications as they prevent us from operating across stages.
+Moreover, several related features (run criteria, states and fixed timesteps) are rather fragile and overly complex.
+This tangled mess needs to be holistically refactored: transferring the ownership of systems to an undivided `Schedule`, simplifying run criteria, taking full advantage of system labels as sets and leveraging exclusive systems to handle complex logic.
+
+## Motivation
+
+The fundamental challenges with the current stage-driven scheduling abstraction are [numerous](https://github.com/bevyengine/bevy/discussions/2801).
+Of particular note:
+
+- Dependencies do not work across stages.
+- States do not work across stages.
+- Plugins can add new, standalone stages.
+- Run criteria (including states and fixed timestep run criteria!) cannot be composed.
+- Run criteria cannot be reused across states or stages.
+- The stack-driven state model is overly elaborate and does not enable enough use cases to warrant its complexity.
+- The architecture for turn-based games and other even slightly unusual patterns is not obvious.
+- We cannot easily implement a builder-style strategy for configuring systems.
+
+Unfortunately, all of these problems are deeply interwoven. Despite our best efforts, fixing them incrementally at either the design or implementation stage is impossible, as it results in myopic architecture choices and terrible technical debt.
+
+## User-facing explanation
+
+This explanation is, in effect, user-facing documentation for the new design.
+In addition to a few new concepts, it throws out much of the current system scheduling design that you may be familiar with.
+The following elements are substantially reworked:
+
+- schedules (simplified, each `World` can store multiple schedules in a `SystemRegistry` resource)

```

Comment:
> - schedules (simplified, each `World` can store multiple schedules in a `SystemRegistry` resource)

The initial PR comment says that schedules are “flattened”. That is a bit ambiguous to me. It could mean one of the following:

1. You cannot nest a schedule inside another.
2. A schedule is linear, not cyclical. It does not have the power to run again by its own. The schedule iteration logic is stored in the app's runner, not in the schedule itself.

It would be nice to have it mentioned here.

Reply:
2. is the correct interpretation. I'll be submitting a PR with updates soon and will clarify this.

Schedules (which are pretty much the executable versions of system sets btw) must be directed acyclic graphs. Any cycles must be encapsulated by an exclusive system.

And yes, there is no internal loop after all the schedule's systems complete.